### PR TITLE
refactor: better error messages and handling

### LINKED ISSUE #604: Better error handling
https://github.com/Azure/aad-pod-identity/pull/590#discussion_r421101856

### PR DESCRIPTION
**Pull request by chewong**
**Reason for Change**:

Logging and returning the error are both considered error handling. We should only choose one method to handle the error but not both.

**Issue Fixed**:

Fixes #604.

**Notes for Reviewers**:

### CODE REVIEW

**Comment by elsesiy** [comment]
@chewong Returning `fmt.Errorf("my string %v", err)` everywhere means you can't test without doing a string comparison. The preferred way in my opinion should be to use `ErrThingGoneWrong` declarations that can be referenced to in a test via `if err == ErrThingGoneWrong`.

**Comment by chewong** [comment]
@elsesiy I agree, being able to compare errors without string comparison in tests will help us drastically increase our test coverage, however, it's probably out of this PR's scope. The purpose of this PR is to avoid logging and returning the same error. Besides, it will require a significant amount of work to refactor the codebase for custom-defined errors but we will definitely keep this in mind when we design the next iteration of aad-pod-identity.

File: pkg/cloudprovider/vm.go [hunk 1]
```diff
@@ -110,13 +107,10 @@ func (c *VMClient) UpdateIdentities(rg, nodeName string, vm compute.VirtualMachi
...
-	if future, err = c.client.Update(ctx, rg, nodeName, compute.VirtualMachineUpdate{
-		Identity: vm.Identity}); err != nil {
-		klog.Errorf("Failed to update VM with error %v", err)
-		return err
+	if future, err = c.client.Update(ctx, rg, nodeName, compute.VirtualMachineUpdate{Identity: vm.Identity}); err != nil {
+		return fmt.Errorf("Failed to update VM with error %v", err)
 	}
 	if err = future.WaitForCompletionRef(ctx, c.client.Client); err != nil {
-		klog.Error(err)
 		return err
```
Comment (aramase) [inline pkg/cloudprovider/vm.go:114]: Wrap this error too so we know the context?

File: pkg/cloudprovider/vmss.go [hunk 2]
```diff
@@ -80,13 +78,10 @@ func (c *VMSSClient) UpdateIdentities(rg, vmssName string, vmssIdentities comput
 		}
 	}()
 
-	if future, err = c.client.Update(ctx, rg, vmssName, compute.VirtualMachineScaleSetUpdate{
-		Identity: vmssIdentities.Identity}); err != nil {
-		klog.Errorf("Failed to update VM with error %v", err)
-		return err
+	if future, err = c.client.Update(ctx, rg, vmssName, compute.VirtualMachineScaleSetUpdate{Identity: vmssIdentities.Identity}); err != nil {
+		return fmt.Errorf("Failed to update VM with error %v", err)
```
Comment (aramase) [inline pkg/cloudprovider/vmss.go:82]: Let's change this to `failed to update identities in VM, error: %v`

File: pkg/mic/mic.go [hunk 3]
```diff
@@ -219,23 +219,21 @@ func NewMICClient(cfg *Config) (*Client, error) {
...
-		return nil, err
+		return nil, fmt.Errorf("failed to create new leader elector, error: %+v", err)
 	}
 	c.leaderElector = leaderElector
 
 	reporter, err := metrics.NewReporter()
 	if err != nil {
-		klog.Errorf("Not able to create New Reporter. Error: %+v", err)
-		return nil, err
+		return nil, fmt.Errorf("failed to create new failed to report metrics, error: %+v", err)
```
Comment (aramase) [inline pkg/mic/mic.go:228]: this log needs update in all packages

File: pkg/mic/mic.go [hunk 4]
```diff
@@ -1131,27 +1128,26 @@ func (c *Client) updateUserMSI(newAssignedIDs map[string]aadpodid.AzureAssignedI
...
 			vmssGroups, getErr := getVMSSGroups(c.NodeClient, nodeRefs)
 			if getErr != nil {
-				klog.Error(getErr)
+				klog.Errorf("failed to get VMSS groups, error: %+v", getErr)
 				continue
 			}
 			inUse, checkErr := c.checkIfInUse(delID, newAssignedIDs, vmssGroups)
 			if checkErr != nil {
-				klog.Error(checkErr)
+				klog.Errorf("failed to check if identity is in used, error: %+v", getErr)
```
Comment (aramase) [inline pkg/mic/mic.go:1136]: s/used/use

File: pkg/filewatcher/filewatcher.go [hunk 5]
```diff
@@ -55,7 +55,7 @@ func (c *client) Start(exit <-chan struct{}) {
 			case <-exit:
 				return
 			case event := <-c.watcher.Events:
-				klog.Infof("Detect file modification: %s", event.String())
+				klog.Infof("detect file modification: %s", event.String())
```
[the remark this answers is not in the document]
Comment (chewong) [reply pkg/filewatcher/filewatcher.go:58]: `event.String()` should output the file name and the operation - https://github.com/fsnotify/fsnotify/blob/master/fsnotify.go#L59-L63

**Pull request merged**